hypershift-operator/controllers/nodepool/inplace.go

```diff
@@ -74,32 +100,362 @@ func (r *NodePoolReconciler) reconcileInPlaceUpgrade(ctx context.Context, hc *hy
+		pod := inPlaceUpgradePod(namespace.Name, node.Name)
+		if result, err := r.CreateOrUpdate(ctx, hostedClusterClient, pod, func() error {
+			return r.reconcileUpgradePod(
+				pod,
+				node.Name,
+				nodePool,
+			)
+		}); err != nil {
+			return fmt.Errorf("failed to reconcile upgrade pod for node %s: %w", node.Name, err)
+		} else {
+			log.Info("Reconciled upgrade pod", "result", result)
+		}
+
+		// Set the actual annotation
+		annotations := map[string]string{
+			DesiredMachineConfigAnnotationKey: targetConfigVersionHash,
+		}
+		if result, err := r.CreateOrUpdate(ctx, hostedClusterClient, node, func() error {
+			return r.reconcileNodeAnnotations(
+				ctx,
+				node,
+				annotations,
+			)
+		}); err != nil {
+			return fmt.Errorf("failed to reconcile node drain annotations: %w", err)
+		} else {
+			log.Info("Reconciled Node drain annotations", "result", result)
+		}
+	}
+	return nil
+}
+
+func (r *NodePoolReconciler) reconcileUpgradePod(pod *corev1.Pod, nodeName string, nodePool *hyperv1.NodePool) error {
+	// TODO (jerzhang): unhardcode some of this
+	configmap := inPlaceUpgradeConfigMap(nodePool, pod.Namespace)
+	pod.Spec.Containers = []corev1.Container{
+		{
+			Name: "machine-config-daemon",
+			// TODO (jerzhang): switch this to MCO image once we have it ready
+			Image: "quay.io/jerzhang/hypershiftdaemon:latest",
+			Command: []string{
+				"/usr/bin/machine-config-daemon",
+			},
+			Args: []string{
+				"start",
+				"--node-name=" + nodeName,
+				"--root-mount=/rootfs",
+				"--kubeconfig=/var/lib/kubelet/kubeconfig",
+				"--desired-configmap=/etc/machine-config-daemon-desired-config",
+			},
+			TerminationMessagePolicy: corev1.TerminationMessageFallbackToLogsOnError,
+			SecurityContext: &corev1.SecurityContext{
+				Privileged: k8sutilspointer.BoolPtr(true),
+			},
+			VolumeMounts: []corev1.VolumeMount{
+				{
+					Name:      "rootfs",
+					MountPath: "/rootfs",
+				},
+				{
+					Name:      "desired-config-mount",
+					MountPath: "/rootfs/etc/machine-config-daemon-desired-config",
+				},
+			},
+		},
+	}
+	pod.Spec.HostNetwork = true
+	pod.Spec.HostPID = true
+	pod.Spec.Tolerations = []corev1.Toleration{
+		{
+			Operator: corev1.TolerationOpExists,
+		},
+	}
+	pod.Spec.NodeSelector = map[string]string{
+		"kubernetes.io/hostname": nodeName,
+	}
+	pod.Spec.Volumes = []corev1.Volume{
+		{
+			Name: "rootfs",
+			VolumeSource: corev1.VolumeSource{
+				HostPath: &corev1.HostPathVolumeSource{
+					Path: "/",
+				},
+			},
+		},
+		{
+			Name: "desired-config-mount",
+			VolumeSource: corev1.VolumeSource{
+				ConfigMap: &corev1.ConfigMapVolumeSource{
+					LocalObjectReference: corev1.LocalObjectReference{
+						Name: configmap.Name,
+					},
+				},
+			},
+		},
+	}
+	pod.Spec.RestartPolicy = corev1.RestartPolicyOnFailure
+
+	return nil
+}
+
+func deleteUpgradePods(ctx context.Context, hostedClusterClient client.Client, nodes []*corev1.Node, nodePool *hyperv1.NodePool) error {
```

Comment:
should this instead be `deleteInPlaceUpgradeNamespace` so it ensures it deletes everything?